crates/bevy_render/src/lib.rs

```diff
@@ -399,6 +358,49 @@ fn extract(main_world: &mut World, render_app: &mut App) {
     main_world.insert_resource(ScratchMainWorld(scratch_world));
 }
 
+/// Initializes the [`FutureRendererResources`] resource.
+fn initialize_render_resources(app: &mut App, settings: WgpuSettings) {
+    let future_renderer_resources_wrapper = Arc::new(Mutex::new(None));
+    app.insert_resource(FutureRendererResources(
+        future_renderer_resources_wrapper.clone(),
+    ));
+
+    let mut system_state: SystemState<Query<&RawHandleWrapper, With<PrimaryWindow>>> =
+        SystemState::new(&mut app.world);
+    let primary_window = system_state.get(&app.world).get_single().ok().cloned();
+
+    bevy_tasks::IoTaskPool::get()
+        .spawn_local(async move {
+            let instance = wgpu::Instance::new(wgpu::InstanceDescriptor {
+                backends: settings.backends.unwrap(),
+                dx12_shader_compiler: settings.dx12_shader_compiler.clone(),
+            });
+            let surface = primary_window.map(|wrapper| unsafe {
+                // SAFETY: Plugins should be set up on the main thread.
```

Comment:
We can't validate this safety invariant purely from the function signature: you could get a `&mut App` on any thread. As a result, this whole function needs to be unsafe, with the safety requirements plumbed up.